pkg/cloud/metadata/k8s.go

```diff
@@ -18,7 +23,39 @@ var DefaultKubernetesAPIClient = func() (kubernetes.Interface, error) {
 	// creates the in-cluster config
 	config, err := rest.InClusterConfig()
 	if err != nil {
-		return nil, err
+		if strings.Contains(err.Error(), "The system cannot find the path specified") {
+			klog.InfoS("InClusterConfig failed to read token file, retrieving file from sandbox mount point")
+			// CONTAINER_SANDBOX_MOUNT_POINT env is set upon container creation in containerd v1.6+
+			// it provides the absolute host path to the container volume.
+			sandboxMountPoint := os.Getenv("CONTAINER_SANDBOX_MOUNT_POINT")
+			if sandboxMountPoint == "" {
+				return nil, fmt.Errorf("CONTAINER_SANDBOX_MOUNT_POINT environment variable is not set")
+			}
+
+			tokenFile := filepath.Join(sandboxMountPoint, "var", "run", "secrets", "kubernetes.io", "serviceaccount", "token")
+			rootCAFile := filepath.Join(sandboxMountPoint, "var", "run", "secrets", "kubernetes.io", "serviceaccount", "ca.crt")
+
+			token, tokenErr := os.ReadFile(tokenFile)
+			if err != nil {
+				return nil, tokenErr
+			}
+
+			tlsClientConfig := rest.TLSClientConfig{}
+			if _, certErr := cert.NewPool(rootCAFile); err != nil {
+				return nil, fmt.Errorf("expected to load root CA config from %s, but got err: %w", rootCAFile, certErr)
+			} else {
+				tlsClientConfig.CAFile = rootCAFile
+			}
+
+			config = &rest.Config{
+				Host:            "https://" + net.JoinHostPort(os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")),
+				TLSClientConfig: tlsClientConfig,
+				BearerToken:     string(token),
+				BearerTokenFile: tokenFile,
```

Comment:
From my understanding, if we specify `BearerTokenFile`, we don't need to also specify `BearerToken` - is that true? Can we remove the `BearerToken` handling?

Reply:
`InClusterConfig` in the `rest` package has both the token and token file in the struct: https://github.com/kubernetes/kubernetes/blob/bc05bef903c063423ac009569974ccbc81138d66/staging/src/k8s.io/client-go/rest/config.go#L511. If token file is specified the value is retrieved periodically and it overrides the `BearerToken` value.

```
BearerTokenFile: Path to a file containing a BearerToken. If set, the contents are periodically read. The last successfully read value takes precedence over BearerToken.
```

Reply:
If upstream does it this way, we'll trust their way /shrug